pkg/costattribution/tracker.go

```diff
@@ -0,0 +1,371 @@
+		}
+		if discardedSampleIncrement > 0 && reason != nil {
+			o.discardedSampleMtx.Lock()
+			if v, ok := o.discardedSample[*reason]; ok {
+				v.Add(discardedSampleIncrement)
+				o.discardedSample[*reason] = v
+			} else {
+				o.discardedSample[*reason] = *atomic.NewFloat64(discardedSampleIncrement)
+			}
+			o.discardedSampleMtx.Unlock()
+		}
+	} else if len(t.observed) < t.maxCardinality*2 && createIfDoesNotExist {
+		// When createIfDoesNotExist is false, it means that the method is called from DecrementActiveSeries, when key doesn't exist we should ignore the call
+		// Otherwise create a new observation for the key
+		t.createNewObservation(key, ts, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement, reason)
+	}
+}
+
+// updateState checks if the tracker has exceeded its max cardinality and updates overflow state if necessary.
+func (t *Tracker) updateState(ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64) {
+	// Transition to overflow mode if maximum cardinality is exceeded.
+	previousState := t.state
+	if t.state == Normal && len(t.observed) > t.maxCardinality {
+		t.state = Overflow
+		// Initialize the overflow counter.
+		t.overflowCounter = &observation{
+			lastUpdate:     atomic.NewInt64(ts),
+			activeSerie:    *atomic.NewFloat64(0),
+			receivedSample: *atomic.NewFloat64(0),
+			totalDiscarded: *atomic.NewFloat64(0),
+		}
+
+		// Aggregate active series from all keys into the overflow counter.
+		for _, o := range t.observed {
+			if o != nil {
+				t.overflowCounter.activeSerie.Add(o.activeSerie.Load())
+			}
+		}
+		t.cooldownUntil = ts + t.cooldownDuration
+	}
+
+	if t.state == Overflow {
+		// if already in overflow mode, update the overflow counter. If it was normal mode, the active series are already applied.
+		if previousState == Overflow && activeSeriesIncrement != 0 {
+			t.overflowCounter.activeSerie.Add(activeSeriesIncrement)
+		}
+		if receivedSampleIncrement > 0 {
+			t.overflowCounter.receivedSample.Add(receivedSampleIncrement)
+		}
+		if discardedSampleIncrement > 0 {
+			t.overflowCounter.totalDiscarded.Add(discardedSampleIncrement)
+		}
+	}
+}
+
+// createNewObservation creates a new observation in the 'observed' map.
+func (t *Tracker) createNewObservation(key []byte, ts int64, activeSeriesIncrement, receivedSampleIncrement, discardedSampleIncrement float64, reason *string) {
+	t.observed[string(key)] = &observation{
+		lastUpdate:         atomic.NewInt64(ts),
+		activeSerie:        *atomic.NewFloat64(activeSeriesIncrement),
+		receivedSample:     *atomic.NewFloat64(receivedSampleIncrement),
+		discardedSample:    map[string]atomic.Float64{},
+		discardedSampleMtx: sync.Mutex{},
+	}
+	if discardedSampleIncrement > 0 && reason != nil {
+		t.observed[string(key)].discardedSampleMtx.Lock()
+		t.observed[string(key)].discardedSample[*reason] = *atomic.NewFloat64(discardedSampleIncrement)
+		t.observed[string(key)].discardedSampleMtx.Unlock()
+	}
+}
```

Comment:
Since we're creating a new observation here, I would pass a `key string` here, allocated once.

I'm not sure if this method isn't allocating 4 strings here (not a big deal, since it's not a hot path, but still..)

Reply:
addressed in 0287bf6cad53e0d4f32fdce6f47f44298efee148